src/proto.rs

```diff
@@ -0,0 +1,484 @@
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Invalid enum variant in field `{field_name}` of message `{message_name}`: `{variant_name}`")]
+    InvalidEnumVariant {
+        field_name: String,
+        message_name: String,
+        variant_name: String,
+    },
+    #[error("Invalid field value in field `{field_name}` of message `{message_name}`: `{value}`")]
+    InvalidFieldValue {
+        field_name: String,
+        message_name: String,
+        value: String,
+    },
+    #[error(
+        "Invalid value in repeated field `{field_name}` of message `{message_name}`: `{value}`"
+    )]
+    InvalidRepeatedValue {
+        field_name: String,
+        message_name: String,
+        value: String,
+    },
+    #[error("Missing field in message `{message_name}`: `{field_name}`")]
+    MissingField {
+        field_name: String,
+        message_name: String,
+    },
+    #[error("Unknown enum tag in field `{field_name}` of message `{message_name}`: `{tag}`")]
+    UnknownEnumTag {
+        field_name: String,
+        message_name: String,
+        tag: i32,
+    },
+}
```

Comment:
Is this file some sort of generated code to do Proto message validation? Or is it written by hand?

Reply:
This is handwritten. I don't know of any codegen tool that can validate proto messages unfortunately.
Some of these error cases, eg `InvalidFieldValue` and `InvalidRepeatedValue`, could not be generated automatically, since they correspond to arbitrary validation logic